quodlibet/util/matcher.py

```diff
@@ -0,0 +1,382 @@
+# Copyright 2021 Joschua Gandert
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+
+from difflib import SequenceMatcher
+from operator import itemgetter
+from typing import Generic, TypeVar, Mapping, Callable, Any, Union, List, Sequence, \
+    Optional, Tuple
+
+
+T = TypeVar('T')
+Real = Union[int, float]
+AttributeGetter = Callable[[T], Any]
+AttributeGetterToWeight = Mapping[AttributeGetter, Real]
+
+SUPPORTED_NUMBER_TYPES = (int, float)
+
+
+class _MatchData:
+    def __init__(self, a_idx, a_value, b_size):
+        self.a_idx = a_idx
+        self.a_value = a_value
+        self.b_idx_to_similarity = [0 for _ in range(b_size)]
+
+        self.best_b_idx = None
+        self.best_b_similarity = float('-inf')
+        self.second_best_b_idx = None
+        self.second_best_b_similarity = float('-inf')
+
+        self.continue_attr_index = 0
+
+        # rarely needed since in a lot of cases there won't be conflicts
+        self._sorted_b_similarity_with_idx_pairs = None
+        self.is_fully_measured = False
+
+    def add_similarity(self, b_idx, similarity_part):
+        self.b_idx_to_similarity[b_idx] += similarity_part
+        b_idx_total_similarity = self.b_idx_to_similarity[b_idx]
+
+        if b_idx_total_similarity > self.best_b_similarity:
+            self.second_best_b_idx = self.best_b_idx
+            self.second_best_b_similarity = self.best_b_similarity
+
+            self.best_b_idx = b_idx
+            self.best_b_similarity = b_idx_total_similarity
+        elif b_idx_total_similarity > self.second_best_b_similarity:
+            self.second_best_b_idx = b_idx
+            self.second_best_b_similarity = b_idx_total_similarity
+
+    def set_to_fully_measured(self):
+        self.is_fully_measured = True
+
+        # We sort so last = best as calling pop() on a list is O(1)
+        # We negate the index here, so that the sorting is correct if the
+        # similarities of two b's are the same (lowest index last)
+        self._sorted_b_similarity_with_idx_pairs = sorted(
+            ((s, -i) for i, s in enumerate(self.b_idx_to_similarity)))
+
+    def replace_best(self):
```

Comment:
Hmm, replace with what? Definitely not following any of this but maybe that's fine

Reply:
I've renamed it to `replace_best_with_next_best`.

Basically we have the following situation: Two lists, a_items and b_items. An a_item is already matched to a b_item (let's say `Song 1` with song 1 (remastered)`). 

Now we measure the similarity of the next a_item and find that it's best match is the exact same b_item. We have a conflict, and we need to find out which a_item is the better match for the b_item. So we compare their similarity (scores) and for the one that loses, we replace it with their second best match.

Then we return to the `while True` loop, but now we have to check if that second best match might also be in conflict... and this repeats until there are no conflicts.